lib/backend/registrybackend/security: Cache authentication tokens

orirawlings opened the pull request:
Fixes #235 

Summary of the changes:
- I noticed that `./lib/dockerregistry/security` seemed to be a more or less copy/pasted, yet unused, version of `./lib/backend/registrybackend/security`. The used version has already drifted so I went ahead and deleted the unused package.
- Rather than building a static credential store for each request, I refactored the code so that the credential store will dynamically return credentials whenever needed. This allows the object to be reused over multiple backend operations.
- I've introduced `./lib/backend/registrybackend/security.NewAuthenticator` to create an object that manages multiple underlying `github.com/docker/distribution/registry/client/auth.TokenHandlers` (one for each repo security scope). This object can be reused over multiple backend operations and the underlying TokenHandlers are reused over requests for the same repo scope.
- I was able to address some comments in the old code that called out some divergent code paths (special cases) for ECR based registries. It turns out that `github.com/docker/distribution/registry/client/auth.Authorizer` can handle multiple AuthorizationHandlers. This allows it to use either BasicAuth or TokenAuth based on the challenge presented by the registry. We just configure the Authorizer with both options every time now, and it will select the correct one on our behalf.
- I took the liberty to break API compatibility of the package during refactoring. I'm not sure what current compatibility guarantees are for the project at this point. If you'd prefer that we preserve compatibility, I can restore compatibility by putting a facade over the new implementation.

We do create an append-only `sync.Map` of token handlers in memory (one entry per repo security scope). Over time this could present as a memory leak if there are an unbounded number of unique repo scopes that a given backend will use. I think in practice, this overhead will be manageable. We can always implement cache expiry/invalidation as a future enhancement.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/kraken?pullRequest=239) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/kraken?pullRequest=239) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/kraken?pullRequest=239) it.</sub>

orirawlings commented:
I'm just working through legal teams at my employer now to get the CLA signed.

orirawlings commented:
Looks like the build is failing with:

yiranwang52 commented:
Don't worry about fossa, that test wouldn't work with forks.

orirawlings commented:
Sorry, things got a bit up-ended at my employer due to COVID-19. I'm still working through our legal process for the CLA, but things are delayed.

orirawlings commented:
Things are unblocked with legal at my employer. CLA is signed now. Sorry for the delay.

evelynl94 commented on lib/backend/registrybackend/blobclient.go:
```diff
@@ -55,41 +56,50 @@ const _manifestquery = "http://%s/v2/%s/manifests/sha256:%s"
 // Stat sends a HEAD request to registry for a blob and returns the blob size.
 func (c *BlobClient) Stat(namespace, name string) (*core.BlobInfo, error) {
-	opt, err := c.config.Security.GetHTTPOption(c.config.Address, namespace)
+	opt, err := c.authenticator.Authenticate(namespace)
```
nice.

evelynl94 commented on lib/backend/registrybackend/security/security.go:
```diff
@@ -36,79 +50,155 @@ type Config struct {
-	switch helper {
+func (c credentialStore) credentialsFromHelper() (string, string) {
+	switch c.config.RemoteCredentialsStore {
+	case "":
```
im assuming this is for public repositories on dockerhub. Could you add a comment?

yiranwang52 commented:
@orirawlings do you mind rebase on master? tests were failing because our self-signed cert expired. that's fixed on master branch.

orirawlings commented:
Thanks for the feedback @evelynl94 and @yiranwang52 

I went ahead and added the requested comment and rebased. I'm getting build failure but I haven't had a chance to investigate yet.

orirawlings commented:
Alright, looks like the build failure was a transient problem resolving dependencies. Retrying the build seems to have fixed things (though the fossa check is failing).

@evelynl94 @yiranwang52 feel free to let me know if you need anything else from me to get this integrated.

yiranwang52 commented:
Thanks for making this fix.
